File: java/src/main/java/org/libpetri/debug/NetEventConverter.java
package org.libpetri.debug;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;

import org.libpetri.core.Token;
import org.libpetri.event.NetEvent;
import org.libpetri.debug.DebugResponse.NetEventInfo;
import org.libpetri.debug.DebugResponse.TokenInfo;

import java.time.Instant;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

/**
 * Converts CTPN NetEvent instances to serializable DebugResponse types.
 */
public final class NetEventConverter {

    /**
     * Shared mapper used to produce the {@code structured} field on {@link TokenInfo}. Separate
     * from {@link NetEventSerializer#SHARED_MAPPER} so we can run with stricter settings here:
     * {@code FAIL_ON_EMPTY_BEANS} disabled keeps non-Jackson-friendly domain objects from
     * exploding the conversion; {@code WRITE_DATES_AS_TIMESTAMPS} disabled gives LLM-facing
     * consumers readable ISO-8601 instants instead of epoch millis.
     */
    private static final ObjectMapper STRUCTURE_MAPPER = new ObjectMapper()
        .registerModule(new JavaTimeModule())
        .disable(SerializationFeature.FAIL_ON_EMPTY_BEANS)
        .disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);

    private NetEventConverter() {}

    /**
     * Attempts to project a token value into a JSON-object-friendly representation (Map / List /
     * JsonNode / primitive) so that downstream serializers emit typed fields rather than a
     * toString string. Returns {@code null} when Jackson throws or would produce an empty node —
     * callers fall back to the existing {@code value} string.
     *
     * <h2>Shape choices</h2>
     * <ul>
     *   <li>{@code null} values return {@code null} (there is nothing structured to show).</li>
     *   <li>Enums are projected to their {@link Enum#name()} — compact and unambiguous.</li>
     *   <li>Strings and boxed primitives pass through unchanged (Jackson emits them as-is).</li>
     *   <li>Records / POJOs round-trip via Jackson; Void tokens never reach this branch.</li>
     * </ul>
     */
    private static Object structuredValue(Object value) {
        if (value == null) return null;
        if (value instanceof Enum<?> e) return e.name();
        if (value instanceof String || value instanceof Number || value instanceof Boolean
                || value instanceof Character) {
            return value;
        }
        try {
            JsonNode tree = STRUCTURE_MAPPER.valueToTree(value);
            // Drop `{}` nodes (opaque beans, private-field only) — the caller's `value` string
            // already carries the toString, adding `"structured": {}` is pure token-budget waste.
            if (tree == null || tree.isNull() || (tree.isObject() && tree.isEmpty())) {
                return null;
            }
            return tree;
        } catch (RuntimeException ex) {
            return null;
        }
    }

    /**
     * Converts a NetEvent to a serializable NetEventInfo.
     *
     * @param event the event to convert
     * @return serializable event info with full token values
     */
    public static NetEventInfo toEventInfo(NetEvent event) {
        return toEventInfo(event, false);
    }

    /**
     * Converts a NetEvent to a serializable NetEventInfo.
     *
     * @param event   the event to convert
     * @param compact if true, token values are omitted (type only)
     * @return serializable event info
     */
    public static NetEventInfo toEventInfo(NetEvent event, boolean compact) {
        return switch (event) {
            case NetEvent.ExecutionStarted e -> new NetEventInfo(
                "ExecutionStarted",
                e.timestamp().toString(),
                null,
                null,
                Map.of(
                    "netName", e.netName(),
                    "executionId", e.executionId()
                )
            );

            case NetEvent.ExecutionCompleted e -> new NetEventInfo(
                "ExecutionCompleted",
                e.timestamp().toString(),
                null,
                null,
                Map.of(
                    "netName", e.netName(),
                    "executionId", e.executionId(),
                    "totalDurationMs", e.totalDuration().toMillis()
                )
            );

            case NetEvent.TransitionEnabled e -> new NetEventInfo(
                "TransitionEnabled",
                e.timestamp().toString(),
                e.transitionName(),
                null,
                Map.of()
            );

            case NetEvent.TransitionClockRestarted e -> new NetEventInfo(
                "TransitionClockRestarted",
                e.timestamp().toString(),
                e.transitionName(),
                null,
                Map.of()
            );

            case NetEvent.TransitionStarted e -> new NetEventInfo(
                "TransitionStarted",
                e.timestamp().toString(),
                e.transitionName(),
                null,
                Map.of(
                    "consumedTokens", tokenList(e.consumedTokens(), compact)
                )
            );

            case NetEvent.TransitionCompleted e -> new NetEventInfo(
                "TransitionCompleted",
                e.timestamp().toString(),
                e.transitionName(),
                null,
                Map.of(
                    "producedTokens", tokenList(e.producedTokens(), compact),
                    "durationMs", e.duration().toMillis()
                )
            );

            case NetEvent.TransitionFailed e -> new NetEventInfo(
                "TransitionFailed",
                e.timestamp().toString(),
                e.transitionName(),
                null,
                Map.of(
                    "errorMessage", e.errorMessage(),
                    "exceptionType", e.exceptionType()
                )
            );

            case NetEvent.TransitionTimedOut e -> new NetEventInfo(
                "TransitionTimedOut",
                e.timestamp().toString(),
                e.transitionName(),
                null,
                Map.of(
                    "deadlineMs", e.deadline().toMillis(),
                    "actualDurationMs", e.actualDuration().toMillis()
                )
            );

            case NetEvent.ActionTimedOut e -> new NetEventInfo(
                "ActionTimedOut",
                e.timestamp().toString(),
                e.transitionName(),
                null,
                Map.of(
                    "timeoutMs", e.timeout().toMillis()
                )
            );

            case NetEvent.TokenAdded e -> new NetEventInfo(
                "TokenAdded",
                e.timestamp().toString(),
                null,
                e.placeName(),
                Map.of(
                    "token", compact ? compactTokenInfo(e.token()) : tokenInfo(e.token())
                )
            );

            case NetEvent.TokenRemoved e -> new NetEventInfo(
                "TokenRemoved",
                e.timestamp().toString(),
                null,
                e.placeName(),
                Map.of(
                    "token", compact ? compactTokenInfo(e.token()) : tokenInfo(e.token())
                )
            );

            case NetEvent.MarkingSnapshot e -> new NetEventInfo(
                "MarkingSnapshot",
                e.timestamp().toString(),
                null,
                null,
                Map.of(
                    "marking", convertMarking(e.marking(), compact)
                )
            );

            case NetEvent.LogMessage e -> {
                var details = new HashMap<String, Object>();
                details.put("loggerName", e.loggerName());
                details.put("level", e.level());
                details.put("message", e.message());
                if (e.throwable() != null) {
                    details.put("throwable", e.throwable());
                }
                if (e.throwableMessage() != null) {
                    details.put("throwableMessage", e.throwableMessage());
                }
                yield new NetEventInfo(
                    "LogMessage",
                    e.timestamp().toString(),
                    e.transitionName(),
                    null,
                    details
                );
            }
        };
    }

    /**
     * Converts a Token to serializable TokenInfo with full value.
     *
     * @param token the token to convert
     * @return token info with full value
     */
    public static TokenInfo tokenInfo(Token<?> token) {
        Object value = token.value();
        String type = value != null ? value.getClass().getSimpleName() : "null";
        String fullValue = value != null ? value.toString() : "null";
        Instant createdAt = token.createdAt();

        return new TokenInfo(
            null,
            type,
            fullValue,
            structuredValue(value),
            createdAt != null ? createdAt.toString() : null
        );
    }

    /**
     * Converts a Token to compact TokenInfo (class name only, no value).
     *
     * @param token the token to convert
     * @return token info with type only
     */
    public static TokenInfo compactTokenInfo(Token<?> token) {
        Object value = token.value();
        String type = value != null ? value.getClass().getSimpleName() : "null";
        Instant createdAt = token.createdAt();
        return new TokenInfo(null, type, null,
            createdAt != null ? createdAt.toString() : null);
    }

    /**
     * Converts a list of tokens to token info list.
     */
    private static List<TokenInfo> tokenList(List<Token<?>> tokens, boolean compact) {
        return tokens.stream()
            .map(compact ? NetEventConverter::compactTokenInfo : NetEventConverter::tokenInfo)
            .toList();
    }

    /**
     * Converts a marking map to serializable form with full token values.
     */
    public static Map<String, List<TokenInfo>> convertMarking(Map<String, List<Token<?>>> marking) {
        return convertMarking(marking, false);
    }

    /**
     * Converts a marking map to serializable form.
     *
     * @param marking the marking to convert
     * @param compact if true, token values are omitted (type only)
     * @return serializable marking
     */
    public static Map<String, List<TokenInfo>> convertMarking(Map<String, List<Token<?>>> marking, boolean compact) {
        Function<Token<?>, TokenInfo> mapper = compact ? NetEventConverter::compactTokenInfo : NetEventConverter::tokenInfo;
        var result = new HashMap<String, List<TokenInfo>>();
        for (var entry : marking.entrySet()) {
            result.put(entry.getKey(), entry.getValue().stream().map(mapper).toList());
        }
        return result;
    }

}

File: typescript/src/debug/net-event-converter.ts
/**
 * Converts CTPN NetEvent instances to serializable DebugResponse types.
 * TypeScript port of Java's NetEventConverter.
 */

import type { Token } from '../core/token.js';
import type { NetEvent } from '../event/net-event.js';
import type { NetEventInfo, TokenInfo } from './debug-response.js';

/** Converts a NetEvent to a serializable NetEventInfo. */
export function toEventInfo(event: NetEvent, compact = false): NetEventInfo {
  switch (event.type) {
    case 'execution-started':
      return {
        type: 'ExecutionStarted',
        timestamp: new Date(event.timestamp).toISOString(),
        transitionName: null,
        placeName: null,
        details: { netName: event.netName, executionId: event.executionId },
      };

    case 'execution-completed':
      return {
        type: 'ExecutionCompleted',
        timestamp: new Date(event.timestamp).toISOString(),
        transitionName: null,
        placeName: null,
        details: {
          netName: event.netName,
          executionId: event.executionId,
          totalDurationMs: event.totalDurationMs,
        },
      };

    case 'transition-enabled':
      return {
        type: 'TransitionEnabled',
        timestamp: new Date(event.timestamp).toISOString(),
        transitionName: event.transitionName,
        placeName: null,
        details: {},
      };

    case 'transition-clock-restarted':
      return {
        type: 'TransitionClockRestarted',
        timestamp: new Date(event.timestamp).toISOString(),
        transitionName: event.transitionName,
        placeName: null,
        details: {},
      };

    case 'transition-started':
      return {
        type: 'TransitionStarted',
        timestamp: new Date(event.timestamp).toISOString(),
        transitionName: event.transitionName,
        placeName: null,
        details: {
          consumedTokens: event.consumedTokens.map(t => compact ? compactTokenInfo(t) : tokenInfo(t)),
        },
      };

    case 'transition-completed':
      return {
        type: 'TransitionCompleted',
        timestamp: new Date(event.timestamp).toISOString(),
        transitionName: event.transitionName,
        placeName: null,
        details: {
          producedTokens: event.producedTokens.map(t => compact ? compactTokenInfo(t) : tokenInfo(t)),
          durationMs: event.durationMs,
        },
      };

    case 'transition-failed':
      return {
        type: 'TransitionFailed',
        timestamp: new Date(event.timestamp).toISOString(),
        transitionName: event.transitionName,
        placeName: null,
        details: {
          errorMessage: event.errorMessage,
          exceptionType: event.exceptionType,
        },
      };

    case 'transition-timed-out':
      return {
        type: 'TransitionTimedOut',
        timestamp: new Date(event.timestamp).toISOString(),
        transitionName: event.transitionName,
        placeName: null,
        details: {
          deadlineMs: event.deadlineMs,
          actualDurationMs: event.actualDurationMs,
        },
      };

    case 'action-timed-out':
      return {
        type: 'ActionTimedOut',
        timestamp: new Date(event.timestamp).toISOString(),
        transitionName: event.transitionName,
        placeName: null,
        details: { timeoutMs: event.timeoutMs },
      };

    case 'token-added':
      return {
        type: 'TokenAdded',
        timestamp: new Date(event.timestamp).toISOString(),
        transitionName: null,
        placeName: event.placeName,
        details: {
          token: compact ? compactTokenInfo(event.token) : tokenInfo(event.token),
        },
      };

    case 'token-removed':
      return {
        type: 'TokenRemoved',
        timestamp: new Date(event.timestamp).toISOString(),
        transitionName: null,
        placeName: event.placeName,
        details: {
          token: compact ? compactTokenInfo(event.token) : tokenInfo(event.token),
        },
      };

    case 'marking-snapshot':
      return {
        type: 'MarkingSnapshot',
        timestamp: new Date(event.timestamp).toISOString(),
        transitionName: null,
        placeName: null,
        details: {
          marking: convertMarking(event.marking, compact),
        },
      };

    case 'log-message': {
      const details: Record<string, unknown> = {
        loggerName: event.logger,
        level: event.level,
        message: event.message,
      };
      if (event.error != null) details['throwable'] = event.error;
      if (event.errorMessage != null) details['throwableMessage'] = event.errorMessage;
      return {
        type: 'LogMessage',
        timestamp: new Date(event.timestamp).toISOString(),
        transitionName: event.transitionName,
        placeName: null,
        details,
      };
    }
  }
}

/**
 * Converts a Token to a serializable {@link TokenInfo}.
 *
 * @remarks
 * The emitted `type` is `value.constructor.name` for objects and `typeof value`
 * for primitives — a *simple name*, not a fully-qualified type identifier.
 * TypeScript has no portable FQN, so cross-language replay from TypeScript
 * archives into a Java reader loses the original type identity (Java's
 * `Class.forName` will fail on simple names) and falls through to the
 * `Token<JsonNode>` graceful-degradation path. The `structured` payload
 * survives intact across languages.
 *
 * The v3 archive body format described in [EVT-025](../../../spec/08-events-observability.md)
 * always emits `structured` alongside `value` so the bundled debug UI keeps
 * rendering while LLM-facing consumers get typed fields. See {@link structuredValue}
 * for the projection rules.
 */
export function tokenInfo(token: Token<unknown>): TokenInfo {
  const value = token.value;
  const type = value != null ? typeof value === 'object' ? value.constructor.name : typeof value : 'null';
  const fullValue = value != null ? String(value) : 'null';
  const info: TokenInfo = {
    id: null,
    type,
    value: fullValue,
    timestamp: new Date(token.createdAt).toISOString(),
  };
  const structured = structuredValue(value);
  return structured === undefined ? info : { ...info, structured };
}

/**
 * Projects a token value into a JSON-friendly representation for the `structured`
 * TokenInfo field. Returns `undefined` when no useful projection exists (null values,
 * opaque objects Jackson-on-the-Java-side would drop); callers omit the field in that
 * case so wire size stays neutral for unstructurable tokens.
 *
 * @remarks
 * Implementation notes:
 * - Primitives and plain objects / arrays pass through untouched (already JSON-safe).
 * - Maps / Sets are projected to a plain shape — debug tooling doesn't need the
 *   prototype identity and JSON consumers can't use it anyway.
 * - Classes with a `toJSON()` method are respected via structured clone (uses the
 *   same path as `JSON.stringify`).
 *
 * Security: uses `JSON.parse(JSON.stringify(...))` which is safe against code
 * execution and prototype pollution. A hostile token value with a custom
 * `toJSON()` override could still return misleading data — archives are a
 * trust boundary (see [EVT-025](../../../spec/08-events-observability.md)).
 */
function structuredValue(value: unknown): unknown | undefined {
  if (value == null) return undefined;
  const t = typeof value;
  if (t === 'string' || t === 'number' || t === 'boolean') return value;
  if (t === 'bigint') return String(value); // bigint → string for JSON-safety
  if (t === 'symbol' || t === 'function') return undefined;
  // Plain arrays / plain objects / objects with toJSON — JSON.parse(JSON.stringify(...))
  // is the cheapest way to drop non-serializable fields and flatten to wire-ready shape.
  try {
    const cloned = JSON.parse(JSON.stringify(value));
    // Skip empty objects — they're opaque beans with no useful structure, just inflating responses.
    if (cloned && typeof cloned === 'object' && !Array.isArray(cloned) && Object.keys(cloned).length === 0) {
      return undefined;
    }
    return cloned;
  } catch {
    return undefined;
  }
}

/** Converts a Token to compact TokenInfo (type only, no value). */
export function compactTokenInfo(token: Token<unknown>): TokenInfo {
  const value = token.value;
  const type = value != null ? typeof value === 'object' ? value.constructor.name : typeof value : 'null';
  return {
    id: null,
    type,
    value: null,
    timestamp: new Date(token.createdAt).toISOString(),
  };
}

/** Converts a marking map to serializable form. */
export function convertMarking(
  marking: ReadonlyMap<string, readonly Token<unknown>[]>,
  compact = false,
): Record<string, readonly TokenInfo[]> {
  const result: Record<string, readonly TokenInfo[]> = {};
  const mapper = compact ? compactTokenInfo : tokenInfo;
  for (const [name, tokens] of marking) {
    result[name] = tokens.map(mapper);
  }
  return result;
}
